### PR TITLE
Haiku: Configuration support

### PR DESCRIPTION
**Pull request by trungnt2910**
Added more support for configuration for Haiku cross-compile builds.

This commit adds recognition for Haiku as a supported platform in various build and configuration files.

For CMake files, this change adds additional guards for non-POSIX headers, functions, and struct members, most notably the `<ucontext.h>` header. (According to the latest POSIX standard, `ucontext_t` should be defined in `<signal.h>`, while `<ucontext.h>` is no longer mandated). `-ldl` is also replaced with `${CMAKE_DL_LIBS}` since Haiku does not have `libdl`.

Part of #55803.

### CODE REVIEW

**Comment by vargaz** [comment]
The mono changes look ok.

File: src/coreclr/.nuget/Directory.Build.props
Comment (AaronRobinsonMSFT) [inline src/coreclr/.nuget/Directory.Build.props:110]: @agocke and @elinor-fung Are we creating new RIDs?

Comment (trungnt2910) [reply src/coreclr/.nuget/Directory.Build.props:110]: Without creating a separate RID for Haiku, builds would fail.

[This](https://github.com/trungnt2910/dotnet-builds/actions/runs/5901050884/job/16010024691) is a result of #90695, which makes certain build components unable to recognize the Haiku RID and therefore cannot resolve some references properly when cross-compiled from an official SDK:

```
/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.cs(6,88): error CS0234: The type or namespace name 'ZLibException' does not exist in the namespace 'System.IO.Compression' (are you missing an assembly reference?) [/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.csproj]
##vso[task.logissue type=error;sourcepath=/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.cs;linenumber=6;columnnumber=88;code=CS0234;](NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'ZLibException' does not exist in the namespace 'System.IO.Compression' (are you missing an assembly reference?)
/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.cs(7,77): error CS0234: The type or namespace name 'CookieVariant' does not exist in the namespace 'System.Net' (are you missing an assembly reference?) [/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.csproj]
##vso[task.logissue type=error;sourcepath=/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.cs;linenumber=7;columnnumber=77;code=CS0234;](NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'CookieVariant' does not exist in the namespace 'System.Net' (are you missing an assembly reference?)
/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.cs(8,77): error CS0234: The type or namespace name 'PathList' does not exist in the namespace 'System.Net' (are you missing an assembly reference?) [/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.csproj]
##vso[task.logissue type=error;sourcepath=/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.cs;linenumber=8;columnnumber=77;code=CS0234;](NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'PathList' does not exist in the namespace 'System.Net' (are you missing an assembly reference?)
  System.Net.Http -> /home/runner/work/dotnet-builds/dotnet-builds/runtime/artifacts/bin/System.Net.Http/Debug/net8.0-haiku/System.Net.Http.dll
  netstandard -> /home/runner/work/dotnet-builds/dotnet-builds/runtime/artifacts/bin/netstandard/Debug/net8.0-haiku/netstandard.dll

Build FAILED.
```

Comment (jkotas) [reply src/coreclr/.nuget/Directory.Build.props:110]: > Are we creating new RIDs?

Yes - for the new base OSes: https://github.com/dotnet/designs/pull/260#issuecomment-1685202292

Comment (elinor-fung) [reply src/coreclr/.nuget/Directory.Build.props:110]: The new base OSes would be added to the trimmed down 'portable' RID graph (https://github.com/dotnet/sdk/pull/34279) that is in sdk now: https://github.com/dotnet/sdk/blob/main/src/Layout/redist/PortableRuntimeIdentifierGraph.json

The idea is that the sdk is the source of truth for the RID graph that gets used for build / nuget restore and the large runtime.json file in runtime is frozen / no longer updated.

Comment (akoeplinger) [reply src/coreclr/.nuget/Directory.Build.props:110]: so the process for adding a new OS now is updating the RID graph in dotnet/sdk and then waiting until we pick it up in dotnet/runtime?

Comment (am11) [reply src/coreclr/.nuget/Directory.Build.props:110]: Yes, the changes in `src/libraries/Microsoft.NETCore.Platforms` are no longer needed for new platforms.

IMO, if we move `PortableRuntimeIdentifierGraph.json` from dotnet/sdk to dotnet/arcade and make that the source-of-truth (in `eng/common` near `eng/common/native/init-distro-rid.sh`), that would provide a better, more reasonable, codeflow.

Comment (trungnt2910) [reply src/coreclr/.nuget/Directory.Build.props:110]: > IMO, if we move `PortableRuntimeIdentifierGraph.json` from dotnet/sdk to dotnet/arcade and make that the source-of-truth (in `eng/common` near `eng/common/native/init-distro-rid.sh`), that would provide a better, more reasonable, codeflow.

That would certainly fix a lot of issues with Haiku builds and remove the current requirement of a Haiku-aware SDK to even build the `dotnet/runtime` repo on Linux.

After #90695 for all Haiku builds I had to manually pass the path to a modified RID graph (See this [commit](https://github.com/trungnt2910/dotnet-builds/commit/363e1db61a80b8ad576cd3ddf0c8d34d3857b634)).

Comment (elinor-fung) [reply src/coreclr/.nuget/Directory.Build.props:110]: > IMO, if we move `PortableRuntimeIdentifierGraph.json` from dotnet/sdk to dotnet/arcade and make that the source-of-truth (in eng/common near eng/common/native/init-distro-rid.sh), that would provide a better, more reasonable, codeflow.

@dsplaisted @ViktorHofer thoughts?

File: src/libraries/System.Data.Odbc/src/System.Data.Odbc.csproj
Comment (ViktorHofer) [inline src/libraries/System.Data.Odbc/src/System.Data.Odbc.csproj:3]: Wow this is really getting out of hand. We should standardize on a common Unix build or even on a common single build and then use platform conditions.

File: src/coreclr/.nuget/Directory.Build.props
Comment (ViktorHofer) [reply src/coreclr/.nuget/Directory.Build.props:110]: cc @tmds

Comment (tmds) [reply src/coreclr/.nuget/Directory.Build.props:110]: The difference between Haiku and Linux is that the Linux has a base rid that is understood by the host and is present in `PortableRuntimeIdentifierGraph.json`.

> if we move PortableRuntimeIdentifierGraph.json from dotnet/sdk to dotnet/arcade and make that the source-of-truth (in eng/common near eng/common/native/init-distro-rid.sh), that would provide a better, more reasonable, codeflow.

I think we may run into issues in the future which will require an updated `PortableRuntimeIdentifierGraph.json` while building runtime (also on Linux). We can solve those when they arise.

In this case (bootstraping a new Unix flavor), imo, we can update `PortableRuntimeIdentifierGraph.json` in the sdk repo as part of the onboarding.

Then a bootstrap SDK for Haiku can be built.

Comment (am11) [reply src/coreclr/.nuget/Directory.Build.props:110]: Either arcade or runtime makes sense for root definitions require to build runtime and its libraries, on which rest of the stack sits. Assuming `PortableRuntimeIdentifierGraph.json` is (an absolutely) necessary artifact, we can derive it from a script or a list out of one of two upstream repos, which anyway require other similar definitions in the porting process.

Comment (tmds) [reply src/coreclr/.nuget/Directory.Build.props:110]: > Assuming PortableRuntimeIdentifierGraph.json is (an absolutely) necessary artifact

As it is now, it is not necessary.

(Honestly, to my surprise) source-build Linux works fine without the (non-portable) rid being built, to be in `PortableRuntimeIdentifierGraph.json`.

If the `runtime` host, and the sdk `PortableRuntimeIdentifierGraph.json` know `haiku` (the same way they know `linux`), that should be enough.

Comment (trungnt2910) [reply src/coreclr/.nuget/Directory.Build.props:110]: I don't know the exact differences between the various different portable or non-portable runtime graphs, but building .NET for Haiku requires a Haiku-aware RID graph.

Otherwise, the build would fail (See [this CI run](https://github.com/trungnt2910/dotnet-builds/actions/runs/5914555471/job/16039990875) for the full logs):
```
/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.cs(6,88): error CS0234: The type or namespace name 'ZLibException' does not exist in the namespace 'System.IO.Compression' (are you missing an assembly reference?) [/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.csproj]
/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.cs(7,77): error CS0234: The type or namespace name 'CookieVariant' does not exist in the namespace 'System.Net' (are you missing an assembly reference?) [/home/runner/work/dotnet-builds/dotnet-builds/runtime/src/libraries/shims/System/src/System.csproj]
```

The problem here is that the build fails to recognize `haiku-x64` as an instance of `unix-x64`, and therefore references generic libraries (`net8.0`) instead of UNIX-specific ones (`net8.0-unix`), therefore missing some UNIX-specific classes.

Comment (ViktorHofer) [reply src/coreclr/.nuget/Directory.Build.props:110]: That's because we have a few libraries that don't offer a Unix compatible implementation and you had to add a `net8.0-haiku` TFM for them. Ideally, that wouldn't be necessary and those libs would just offer a Unix compatible implementation with runtime checks. From the changes in this PR it doesn't look like there are a lot of these. System.IO.Ports is a great example of a library that would benefit from runtime over build-time checks.

Comment (tmds) [reply src/coreclr/.nuget/Directory.Build.props:110]: > I don't know the exact differences between the various different portable or non-portable runtime graphs

The legacy graph is called the non-portable graph. It is stored in the runtime repo. It is bloated with non-portable rids. It will probably disappear in .NET 9.

With .NET 8, there's a new graph, named `PortableRuntimeIdentifierGraph.json`. It is stored in the sdk repo. We still add a non-portable rid to it during source-build. .NET 8 uses this one by default.

> Otherwise, the build would fail (See [this CI run](https://github.com/trungnt2910/dotnet-builds/actions/runs/5914555471/job/16039990875) for the full logs):

For this we need to patch portable graph. And until it moves somewhere else, I think you can make a PR against the sdk repo that adds `haiku` to it.

You can try this change locally by patching the file in the SDK you are using for the `runtime` build.

When it is added to the sdk repo, it will (after some time) become part of the SDK used by the runtime repo, and the CI leg may turn green.

Although it sounds slow, it may happen faster than moving the rid graph around.
It took until .NET 7 until source-build was capable of building on unknown Linux rids. :sweat_smile:

Comment (tmds) [reply src/coreclr/.nuget/Directory.Build.props:110]: > So, what you're suggesting is that I should revert the changes to the non-portable RID in this PR, and then open a separate PR to dotnet/sdk to add Haiku support to the portable RID graph?

Yes. The graph that is used for .NET8+ is currently in the sdk repo, so you can make a PR against that.
If it moves some place else, your changes will still be included.

Comment (dsplaisted) [reply src/coreclr/.nuget/Directory.Build.props:110]: I am fine with moving the portable RID graph back to the runtime repo (or dotnet/arcade, but to me runtime makes more sense).

Comment (trungnt2910) [reply src/coreclr/.nuget/Directory.Build.props:110]: > Yes. The graph that is used for .NET8+ is currently in the sdk repo, so you can make a PR against that.

https://github.com/dotnet/sdk/pull/35231

Made a PR for it.

After that is merged, I will remove the changes to `src/libraries/Microsoft.NETCore.Platforms` on this branch.

Comment (ViktorHofer) [reply src/coreclr/.nuget/Directory.Build.props:110]: This will unblock source building runtime for haiku but you won't be able to add new "haiku" Target Platform Identifiers without source build conditions until the updated portable RID graph is available in the non source built SDK.

We should probably just move the portable RID graph into runtime and keep the Platforms package as a transport package for the sdk repo.

Comment (ViktorHofer) [reply src/coreclr/.nuget/Directory.Build.props:110]: Same as with the old non-portable RID graph. We would keep both graphs in runtime and have a build task that allows us to add a RID to the graph on-the-fly. That updated portable RID graph would then flow together with the old non-portable RID graph into dotnet/sdk as a transport package. dotnet/sdk would then pick both graphs up and add them to the shipping SDK folder.

Comment (ViktorHofer) [reply src/coreclr/.nuget/Directory.Build.props:110]: Submitted https://github.com/dotnet/runtime/pull/92211 targeting main to bring the portable RID graph over into runtime.

Comment (trungnt2910) [reply src/coreclr/.nuget/Directory.Build.props:110]: I've rebased this branch to match the upstream main branch (which contains #92211).

I've also removed the version-specific Haiku RIDs from `runtime.json`. Should I remove all Haiku RIDs from that graph altogether though?

Comment (tmds) [reply src/coreclr/.nuget/Directory.Build.props:110]: The old graph is probably staying around for .NET 9 (https://github.com/dotnet/runtime/pull/92211#issuecomment-1725943953) so it's best to keep `haiku` in it.

Comment (trungnt2910) [reply src/coreclr/.nuget/Directory.Build.props:110]: > it's best to keep haiku in it.

Just `haiku`, or also the versioned one (`haiku.1` for Haiku R1)?

Comment (tmds) [reply src/coreclr/.nuget/Directory.Build.props:110]: Just `haiku`. During a non-portable build, `haiku.1` will be automatically added to both graphs.

File: src/libraries/Microsoft.NETCore.Platforms/src/runtime.json
Comment (am11) [inline src/libraries/Microsoft.NETCore.Platforms/src/runtime.json:1731]: Update [PortableRuntimeIdentifierGraph.json](https://github.com/dotnet/runtime/commit/3b9b4fd5fcd3ff9ed0ea1622ff42f98a1068316b#diff-588ddddd3493c7445abeb05129392c25b9d03a30133085f6ec4b78d2b1adbe49) as well, which was recently moved in the repo?

Comment (trungnt2910) [reply src/libraries/Microsoft.NETCore.Platforms/src/runtime.json:1731]: > Update [PortableRuntimeIdentifierGraph.json](https://github.com/dotnet/runtime/commit/3b9b4fd5fcd3ff9ed0ea1622ff42f98a1068316b#diff-588ddddd3493c7445abeb05129392c25b9d03a30133085f6ec4b78d2b1adbe49) as well, which was recently moved in the repo?

Probably not, because before it moved here, https://github.com/dotnet/sdk/pull/35231 containing the changes for Haiku in the portable graph has already been merged.

Comment (trungnt2910) [reply src/libraries/Microsoft.NETCore.Platforms/src/runtime.json:1731]: See also: https://github.com/dotnet/runtime/blob/bbe5c66c50201816c14149fd3f259db86bc1f493/src/libraries/Microsoft.NETCore.Platforms/src/PortableRuntimeIdentifierGraph.json#L67-L77

File: src/coreclr/.nuget/Directory.Build.props
Comment (trungnt2910) [reply src/coreclr/.nuget/Directory.Build.props:110]: This PR still keeps `haiku`, so I guess things should work.

**Review by am11 — APPROVED** [review]
LGTM! :shipit:

File: src/libraries/System.Data.Odbc/src/System.Data.Odbc.csproj
Comment (trungnt2910) [reply src/libraries/System.Data.Odbc/src/System.Data.Odbc.csproj]: Rebased on `main` and removed the changed adding `haiku` to this list in favor of a unified `unix` build.

File: src/tests/Directory.Build.props
Comment (trungnt2910) [reply src/tests/Directory.Build.props]: Done. Let's see if it breaks any builds.

**Review by jkotas — APPROVED** [review]
LGTM. Any concerns with getting this PR merged?

**Review by ViktorHofer — APPROVED** [review]
The libraries changes look good to me. I would like to change the four touched projects (Net.Http, Net.NetworkInformation, Net.Security and Serialization.Formatters) in the future so that we don't have to introduce new platform TFMs.

**Comment by jkotas** [comment]
@trungnt2910 Could you please resolve the conflict?

**Comment by trungnt2910** [comment]
> @trungnt2910 Could you please resolve the conflict?

Done.

**Comment by jkotas** [comment]
@trungnt2910 Thank you!

**Pull request merged**
